### PR TITLE
Use a simplified magic-string.

### PR DESCRIPTION
**Pull request by jdalton**
This is a start of a simplified magic-string.

### CODE REVIEW

File: lib/magic-string.js [hunk 1]
```diff
@@ -0,0 +1,142 @@
...
+    this.start = start;
+    this.end = end;
+    this.original = content;
+    this.intro = "";
+    this.outro = "";
+    this.content = content;
+    this.next = null;
+  }
+
+  appendLeft(content) {
```
Comment (jdalton) [inline lib/magic-string.js]: for classes do you prefer `appendLeft (` or `appendLeft(` (space, no space)

File: lib/magic-string.js [hunk 2]
```diff
@@ -0,0 +1,142 @@
...
+
+    while (true) {
+      if (chunk.contains(index)) {
+        this._splitChunk(chunk, index);
+        return;
+      }
+      chunk = searchForward ?
+        this.byStart[chunk.end] :
+        this.byEnd[chunk.start];
+    }
```
Comment (jdalton) [inline lib/magic-string.js]: What's your ternary style preference?

File: lib/magic-string.js (hunk 1 above)
Comment (benjamn) [inline lib/magic-string.js]: No space after a function or method name. I only put a space before the `(` if it's an anonymous function.

File: lib/magic-string.js (hunk 2 above)
Comment (benjamn) [inline lib/magic-string.js]:
```js
chunk = searchForward
  ? this.byStart[chunk.end]
  : this.byEnd[chunk.start];
```
please.

File: lib/magic-string.js [hunk 3]
```diff
@@ -0,0 +1,142 @@
...
+    this.byStart[index] = newChunk;
+    this.byEnd[newChunk.end] = newChunk;
+    this.lastSearchedChunk = chunk;
+  }
+
+  toString() {
+    let str = this.intro;
+    let chunk = this.firstChunk;
+    while (chunk) {
+      str += chunk.toString();
```
Comment (benjamn) [inline lib/magic-string.js]: Maybe build up an array of chunks and then `.join("")` at the end?

**Comment by jdalton** [comment]
Ya, I created this build by taking the original magic-string.cjs.js and removed stuff, checked if unit tests failed, then removed more. Having expected output tests though is a good thing. I don't think we need to compare to magic-string proper but have a baseline (which could be this) of what we think is OK so it can catch bugs that may slip in. (I do this with babel-plugin-lodash.

**Comment by jdalton** [comment]
If it helps we can add tests based on the magic-string package output (before this PR) then just rebase and see if they pass.

File: lib/magic-string.js (hunk 3 above)
Comment (jdalton) [inline lib/magic-string.js]: That optimization has gone out of vogue (no longer needed in modern engines).
The loop as is, is the code magic-string uses.

Comment (jdalton) [inline lib/magic-string.js]: Engines will defer flattening a string until it's necessary. Using `Array#join` forces an immediate flattening.

Comment (benjamn) [inline lib/magic-string.js]: Interesting!

**Comment by jdalton** [comment]
Added a bunch of tests using `"magic-string/dist/magic-string.cjs.js"` results as the expected output.

**Comment by benjamn** [comment]
Awesome! Any feelings about whether we should bump the version to `0.6.11` or `0.7.0`?

**Comment by jdalton** [comment]
This prolly fits a `0.7.0`.

**Pull request merged**